text/0000-system-receptionist.md

```diff
@@ -0,0 +1,255 @@
+
+
+### Implementation
+
+
+**Creating a system**
+
+The `createSystem(rootMachine, options?)` function will create an `ActorRef` whose behavior is defined by the `rootMachine`, which is a normal machine created from `createMachine(...)`. The purpose of the system is to provide `{ system }` to all machines (including the root) spawned in this system, which contains system-wide resources.
+
+```ts
+import { createSystem, createMachine } from 'xstate';
+
+const rootMachine = createMachine({
+  // ...
+});
+
+// Creating a system
+const system = createSystem(rootMachine);
+
+// Starting a system
+system.start();
+```
+
+**Registering to the receptionist**
+
+The `register(key)` action is made available from `'xstate/actions'`, which will register the executing actor to the system-wide registry under that `key`. Multiple actors can be registered under the same `key` by default, and it should be assumed that each `key` in the registry has 0 or more actor refs.
+
+```ts
+import { register } from 'xstate/actions';
+
+const todoMachine = createMachine({
+  // Register actor to the 'todo' key
+  entry: register('todo'),
+  // ...
+});
+```
+
+This should be roughly equivalent to:
+```ts
+// PSEUDOCODE
+entry: (ctx, e, { system, self }) => {
+  system.register('todo', self);
+}
+```
+
+**Getting listings**
+
+The `system` value is made available in the 3rd argument to action/guard/invoke creators, which contains the `system.receptionist` that extends the `ActorRef` interface.
+
+Listing all `ActorRef` objects registered under a key is done via `system.receptionist.list(key)`, which synchronously returns an array of `ActorRef` objects.
+
+```ts
+const todosMachine = createMachine({
+  // ...
+  on: {
+    'todos.logAll': {
+      actions: (ctx, e, { system }) => {
+        // Lists all todo ActorRefs
+        console.log(system.receptionist.list('todo'));
+      }
+    },
+    'todos.completeAll': {
+      actions: pure((ctx, e, { system }) => {
+        return system.receptionist.list('todo').map(todoRef => {
+          return sendTo(todoRef, { type: 'complete' });
+        });
+      });
+    }
+  }
+});
+```
+
+This should be roughly equivalent to:
+```ts
+// PSEUDOCODE
+// system.receptionist.list('todo');
+system.receptionist.getSnapshot().registry.get('todo') ?? []; // => ActorRef[]
+```
+
+**Single listing**
+
+A common use-case of the receptionist will be to register and find a single `ActorRef` registered to a given `key`. This can be done via `system.registry.find(key)`, which will return `ActorRef | null`.
+
+```ts
+const todoMachine = createMachine({
+  // ...
+  on: {
+    complete: {
+      // `system.registry.find(key)` returns 0 or 1 ActorRef objects
+      // TODO: ensure `sendTo(...)` handles null targets (dead-letter queue?)
```

Comment:
how often does such a swallowing behavior implicitly allows bugs and how often this is actually what a developer wants?